dev/bots/custom_rules/debug_assert.dart

```diff
@@ -0,0 +1,382 @@
+                                     || classMember.metadata.any(_isDebugAssertAnnotationElement)
+                                     || classMember.enclosingElement.metadata.any(_isDebugAssertAnnotationElement);
+
+    bool isAnnotated = isClassMemberAnnotated;
+    for (final InterfaceType supertype in enclosingElement.allSupertypes) {
+      if (supertype.element.library.isInSdk) {
+        continue;
+      }
+      final ExecutableElement? superImpl = switch (classMember) {
+        MethodElement(:final String name)                           => supertype.getMethod(name),
+        PropertyAccessorElement(:final String name, isGetter: true) => supertype.getGetter(name),
+        PropertyAccessorElement(:final String name, isSetter: true) => supertype.getSetter(name),
+        _ => throw StateError('not reachable $classMember(${classMember.runtimeType})'),
+      };
+      if (superImpl == null) {
+        continue;
+      }
+      final bool isSuperImplAnnotated = _overriddableClassMemberHasDebugAnnotation(superImpl, supertype.element);
+      if (isClassMemberAnnotated && !isSuperImplAnnotated) {
+        incorrectAnnotations.add(_IncorrectAnnotationError(superImpl, classMember));
+      }
+      isAnnotated |= isSuperImplAnnotated;
+    }
+
+    return overridableMemberLookup[classMember] = isAnnotated;
+  }
+
+  bool _syntheticConstructorHasDebugAnnotation(ConstructorElement constructorElement) {
+    assert(constructorElement.isSynthetic);
+    assert(constructorElement.isDefaultConstructor);
+    if (constructorElement.library.isInSdk) {
+      return false;
+    }
+
+    assert(!constructorElement.isFactory);
+    assert(constructorElement.isDefaultConstructor);
+
+    // Subclasses may "inherit" default constructors from the superclass. Since
+    // constructors can't be invoked by the class members (unlike methods that
+    // can have "bad annotations"). Defer to the first non-synthetic default
+    // constructor in the class hierarchy.
+    final ConstructorElement? superConstructor = constructorElement.enclosingElement.thisType
+      .superclass?.element.constructors
+      .firstWhereOrNull((ConstructorElement constructor) => constructor.isDefaultConstructor);
+    return superConstructor != null && _isDebugOnlyExecutableElement(superConstructor);
+  }
+
+  bool _isDebugOnlyExecutableElement(Element element) {
+    final LibraryElement? lib = element.library;
+    if (lib == null || lib.isInSdk) {
+      // The assert is for framework code only. Dart sdk symbols won't be annotated.
+      return false;
+    }
+
+    switch (element) {
+      // Constructors are static in nature so there won't be any "bad annotations"
+      // (see the non-static ExecutableElement case), but a default synthetic
+      // constructor should defer to its superclass since it's not explictly
+      // defined.
+      case ConstructorElement(isSynthetic: true):
```

Comment:
I'm not sure why the synthetic nature makes a difference. Shouldn't this code be reported?

```dart
class C {
  @debugAssert
  C();
}

class D extends C {
  D(); // Implicitly calls C().
}

Reply:
See https://github.com/flutter/flutter/pull/130101#discussion_r1297865247 

Also that code will be reported I think? Neither `C.new` nor `D.new` is synthetic here?